pkg/conduit/runtime.go

```diff
@@ -191,7 +191,7 @@ func newServices(
 // Run initializes all of Conduit's underlying services and starts the GRPC and
 // HTTP APIs. This function blocks until the supplied context is cancelled or
 // one of the services experiences a fatal error.
-func (r *Runtime) Run(ctx context.Context) (err error) {
+func (r *Runtime) Run(ctx context.Context, cancel context.CancelFunc) (err error) {
 	t, ctx := tomb.WithContext(ctx)
```

Comment:
The first thing we do is instantiate a tomb, so instead of requiring the caller to send us a cancel function we can simply kill the tomb with `t.Kill(reason)`, that will cancel the context. A nice thing is also that we can supply the error that caused the death of the tomb, which is then returned by `Run` and logged by Conduit as the cause for why it stopped.

Reply:
That was actually my first option. Then I realized that `tomb.WithContext` returns a new context, so the original one itself (given to `Run()`) cannot be canceled. But _now_ I see we're actually overwriting the input context. I guess that's done so that whoever is using the "main Conduit context" also gets notified of the cancellation when this tomb is killed. I had a concern that `Run` can mutate `ctx` in a way in which the original cancellation is lost (e.g. `tomb.WithContext(context.Background())`). But it looks like there's a test for this, so I think your suggestion makes a lot of sense.